{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Import external modules used in the paper\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import gamma, chi2\n",
    "import itertools\n",
    "# This is the only non-standard module used here. We only use it to implement false-discovery rate correction\n",
    "from statsmodels.sandbox.stats.multicomp import fdrcorrection0 as fdr "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### This code takes in a data set that consists of p-values for the differential expression of each\n",
    "### orthology group in each organism undergoing regeneration. The bulk of those code is there to process\n",
    "### the data and convert it into a Pandas data frame that is amenable to efficient statistical analysis.\n",
    "### The main piece of analysis in this cell performs p-value aggregation using the Lancaster method.\n",
    "### This cell initializes the library of custom functions that are used in later cells\n",
    "\n",
    "\n",
    "# This function takes in the text file containing RNA-seq data that contains (log) read counts and p-values for time series from each gene in each organism we use\n",
    "# for the analysis. \n",
    "def read_file(folder, file):\n",
    "    \"\"\"\n",
    "    This function takes in a text file containg log_2 read counts and p-values from RNA-seq time series and converts it to a Pandas data frame.\n",
    "    This function is intended to be fairly generic, while the internal function read_file() is meant to be specialized to the particular format of the input data.\n",
    "    Args:\n",
    "        folder (string): Folder where the text file is located and where we will store saved data.\n",
    "        file (string): Text file with RNA-seq data.\n",
    "    Returns:\n",
    "        dict: Dictionary where keys corresponding to the differnet organisms we compare, and values consisiting of (gene, ortholog group, read count, p-value) data.\n",
    "        \n",
    "    \"\"\"\n",
    "    data = {} # initialize output. This is a dictionary that will be modeled after a json-like structure. \n",
    "    with open(folder + file) as f:\n",
    "        content = f.readlines()\n",
    "        \n",
    "        for line in content:\n",
    "            row = line.rstrip().split() # just some simple formating and removal of whitespace\n",
    "            if row: # this block just checks to make sure the row isn't empty, and gets rid of some elements that aren't necessary for the purposes of this code.\n",
    "                try:\n",
    "                    del row[1]\n",
    "                except Exception as e:\n",
    "                    print(row)\n",
    "                org = row[1]\n",
    "                \n",
    "                if org not in data.keys(): # initialize new entry when we get to a new organism\n",
    "                    data[org] = []\n",
    "                    \n",
    "                reads = [float(el) for el in row[3::2]] # log_2 read counts for the gene\n",
    "                pvals =  [float(el) for el in row[4::2]] # p-values for the gene\n",
    "                \n",
    "                if len(pvals) > 0:\n",
    "                    entry = {'gene': row[0], 'ogroup': row[2], 'reads':reads, 'pvals':pvals}\n",
    "                    data[org].append(entry)\n",
    "    return data\n",
    "    \n",
    "\n",
    "def file2df(folder,file):\n",
    "    \"\"\"\n",
    "    This function takes in a text file containg log_2 read counts and p-values from RNA-seq time series and converts it to a Pandas data frame.\n",
    "    This function is intended to be fairly generic, while the internal function read_file() is meant to be specialized to the particular format of the input data.\n",
    "    Args:\n",
    "        folder (string): Folder where the text file is located and where we will store saved data.\n",
    "        file (string): Text file with RNA-seq data.\n",
    "    Returns:\n",
    "        DataFrame: Pandas data frame on which we perform statistical analysis\n",
    "        \n",
    "    \"\"\"\n",
    "    data = read_file(folder,file) \n",
    "    data_agg = []\n",
    "    \n",
    "    for key in data.keys():\n",
    "        for row in data[key]:\n",
    "            if row['pvals']:\n",
    "                row['organism'] = key\n",
    "                data_agg.append(row)\n",
    "        print(len(data_agg))\n",
    "    df = pd.DataFrame(data_agg)\n",
    "    return df\n",
    "\n",
    "\n",
    "def lancaster(row):\n",
    "    \"\"\"\n",
    "    This is a Python implementation of the Lancaster method for p-value aggregation. This implementation is based on the method described in\n",
    "    Gene-level differential analysis at transcript-level resolution, Yi et al. 2018, Genome Biology.\n",
    "    Args:\n",
    "        row (pd.Series): a row of a data frame consisting of a list of p-values and log_2 read counts\n",
    "    Returns:\n",
    "        p (float): aggregated p-value\n",
    "    \"\"\"\n",
    "    tau = 0\n",
    "    pvals = row['pvals']\n",
    "    reads = [2**r for r in row['reads']]\n",
    "    t = [gamma.isf(max(p,tau),w/2, scale=2) for p, w in zip(pvals,reads)]\n",
    "    t = sum(t)\n",
    "    p = chi2.sf(t,sum(reads))\n",
    "    return p\n",
    "\n",
    "\n",
    "def overlap_subsets(Ogroup_agg):\n",
    "    \"\"\"\n",
    "    This function takes in a data frame consisting of ortholog groups from each organism with their associated lancaster p-values.\n",
    "    It then adjusts the p-values to account for false-discovery rate corrections. It then uses these final p-values to determine which orthology groups are present\n",
    "    across each possible subset of organisms for which we have data.\n",
    "    Args:\n",
    "        Ogroup_agg (pd.DataFrame): Data frame that comes from the\n",
    "    Returns:\n",
    "         Data Frame: Data frame consisting of each possible combination of organisms and the ortholog groups that pass significance test within the subset\n",
    "    \"\"\"\n",
    "    og_lists = {}\n",
    "    agg_list = []\n",
    "    # false-discovery rate correction applied to list of lancaster p-values computed for ortholog groups for each organism. the function fdr() returns the orthology groups which\n",
    "    # have adjusted p-value > 0.05.\n",
    "    for name, group in Ogroup_agg.groupby('organism'): \n",
    "        padj = fdr(group.p_lanc,alpha = .05) \n",
    "        og_l = list(group.ogroup.loc[padj[0]]) # list of ortholog groups for a given organism that pass significance test\n",
    "        og_lists[name] = og_l\n",
    "        agg_list += og_l\n",
    "    agg_list = np.unique(agg_list)\n",
    "\n",
    "    overlap = {}\n",
    "    for og in agg_list: # iterate over every ortholog group and track which organisms \n",
    "        overlap[og] = []\n",
    "        for key in og_lists:\n",
    "            if og in og_lists[key]:\n",
    "                overlap[og].append(key)\n",
    "\n",
    "    ol_subsets = dict()\n",
    "    orgs = Ogroup_agg.organism.unique()\n",
    "    \n",
    "    for i in range(orgs.size): # iterate over every possible subset of organisms and compile list of ortholog groups that pass significance threshold in each organism in a given subset\n",
    "        for el in itertools.combinations(orgs, i+1): \n",
    "            ol_subsets[str(list(el))] = pd.Series([og for og in agg_list if set(el).issubset(overlap[og])])\n",
    "            \n",
    "    return pd.DataFrame.from_dict(ol_subsets)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15774\n",
      "64952\n",
      "115533\n",
      "164238\n",
      "189106\n",
      "222860\n",
      "238745\n"
     ]
    }
   ],
   "source": [
    "### This cells runs the analysis that takes in a text file with RNA-seq data and outputs a Pandas data frame and computes aggregated p-values for each orthology group. \n",
    "\n",
    "# file = '1_Annotated_Pvalue_Matrix.6-2-18.txt'\n",
    "# folder = '06-07-18_Data/'\n",
    "file = '1_Annotated_Pvalue_Matrix.HeatShock.9-21-2020.txt'\n",
    "folder = '09-30-20_Data/'\n",
    "df = file2df(folder, file)\n",
    "df.to_csv(folder + 'pcounts.csv',index=False) #saves the data frame version of the text file\n",
    "\n",
    "# Some reorganization of the data frame so that we aggregate all paralogs from the same ortholog group. This implicitly assumes\n",
    "# that these p-values can be treated as independent test of the differnetial expression hypothesis.\n",
    "Ogroup_agg = df.groupby(['organism','ogroup'],as_index=True)['pvals','reads'].sum() \n",
    " \n",
    "Ogroup_agg['p_lanc'] = Ogroup_agg.agg(lancaster,axis=1) # apply the Lancaster method to each orthology group. This step can take a few minutes!\n",
    "Ogroup_agg.to_csv(folder + 'lanc.csv') # saves the aggregated data frame with the lancaster p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = pd.read_csv(folder + 'lanc.csv')\n",
    "overlap_df = overlap_subsets(A)\n",
    "overlap_df.to_csv(folder + 'ol_subsets.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
