###### RNAseq_pValueAggregation
# Assessing homology through comparative RNAseq


#### The objective of this analysis is to search for potential homology in biological processes using existing RNAseq datasets, correcting for differences in experimental design.

The analysis combines differential expression statistics for individuals genes (edgeR) with orthology analysis (orthofinder), and then tests whether the distribution of p-values for differential expression significance of the genes within a conserved ortholog group(COG) are likely to have been generated from a random distribution (0,1). If a COG is involved in the relevant biological process across all organisms tested (the distribution of p-values from the genes within that COG are skewed towards significance (1)), it would suggest a conserved function of that COG within the relevant biological process.

## Scripts
#### This package contains three scripts:
- comp_RNAseq.sh
- generate_pValue_input.sh
- pval_aggregation.py


&nbsp;&nbsp;&nbsp;&nbsp;**comp_RNAseq.sh** performs:

  0) Downloads data for heat shock and regeneration experiments used in this paper
  1) RSEM-SE mapping to genomes/transcriptomes to find identities of reads in datasets
  2) Differential Gene Expression analysis with Trinity:edgeR to identify genes of interest in each experiment
		- run_DE_analysis.pl: identify differentially expressed transcripts
		- analyze_diff_expr.pl: extract the DE transcripts and cluster them according to their patterns across the conditions
		  - output is clusters of genes expressed in similar patterns under heat shock for each animal
  3) Transdecoder translater the reads into protein sequences so orthology can be determined on the basis of protein structure (function) rather than nucleotide sequence
  4) OrthoFinder clusters transcripts by homology, performed on the files of condition-clustered transcript lists
		- performed in two steps: blast search on hpc, orthofinder on completed blast searches locally
		  - output is orthologue groups for transcripts differentially expressed within the treatments across all species

     
&nbsp;&nbsp;&nbsp;&nbsp;**generate_pValue_input.sh** generates an input file for pval_aggregation.py combining differential expression data from edgeR and orthogroups from orthofinder.

&nbsp;&nbsp;&nbsp;&nbsp;**pval_aggregation.py** uses the Lancaster Aggregation method to test whether the differential expression p-values of genes contained within a COG are likely to have been generated from a uniform distribution.

This aggregation method from Yi et al. takes advantage of the fact that many independent p-values generated by the null hypothesis should follow a uniform distribution on the interval (0,1). Consequently, we can test the uniformity of the set of p-values to determine their likelihood of being generated from the null hypothesis, which in our case corresponds to the COG not being differentially expressed. This approach allows us to make statistically honest comparisons of differential gene expression between diverse studies. 

## Input files 
The data for our analyses are contained within this repo and can be downloaded within [comp_RNAseq.sh](https://github.com/nsierra1/RNAseq_pValueAggregation/blob/main/comp_RNAseq.sh). To use this process with new datasets, the following input files are required:

- **accession_spp.txt**: the SRR accession values (or other accession codes) for each species must be stored in a txt file
- **gunzip_SRR.sh**: the descriptive names (experimental condition) for each SRR can be specified during the unzip process 
&nbsp;&nbsp;&nbsp;&nbsp;(ex: raw/SRR6432693_1.fastq.gz > 23C_6h_1.fastq denotes the temp(23C) for the 6hr condition, replicate 1)
- **genome_accession.sh**: the url for fasta format genome or transcriptome files. These can also be unzipped with descriptive names.
- **spp_DEcomp.txt**: filenames for edgeR output comparing adjacent time points for input to generate_pValue_input.sh
&nbsp;&nbsp;&nbsp;&nbsp;(ex: comparing time0 to time1 and time 1 to time 2, but not time0 to time2)

&nbsp;&nbsp;&nbsp;&nbsp;For examples on formatting these files please see the [supporting files](https://github.com/nsierra1/RNAseq_pValueAggregation/tree/main/supporting_files) from our analysis.

